File: 2016/Day22/main.py
from dataclasses import dataclass
from queue import Queue
import re
import sys

@dataclass(frozen=True)
class Node:
    x: int
    y: int
    size: int
    used: int
    avail: int

@dataclass(frozen=True)
class State:
    empty_x: int
    empty_y: int
    goal_x: int
    goal_y: int
    steps: int

def print_grid(nodes: list[Node], max_x: int, max_y: int, empty_x: int, empty_y: int, goal_x: int):
    grid = [['.' for _ in range(max_x + 1)] for _ in range(max_y + 1)]

    for node in nodes:
        if node.used > 100:
            grid[node.y][node.x] = '#'
    
    grid[empty_y][empty_x] = '_'
    grid[0][0] = 'O'
    grid[0][goal_x] = 'G'

    print('Grid visualization:')
    for y in range(max_y + 1):
        print(' '.join(grid[y]))
    
def solve_part2(nodes: list[Node]) -> int:
    max_x = max(node.x for node in nodes)
    max_y = max(node.y for node in nodes)

    empty_x, empty_y = -1, -1
    for node in nodes:
        if node.used == 0:
            empty_x, empty_y = node.x, node.y
            break
    
    goal_x, goal_y = max_x, 0

    viable_nodes = [[True for _ in range(max_x + 1)] for _ in range(max_y + 1)]
    for node in nodes:
        if node.used > 100:
            viable_nodes[node.y][node.x] = False
    
    print_grid(nodes, max_x, max_y, empty_x, empty_y, goal_x)

    q: Queue[State] = Queue()
    visited: set[State] = set()

    start = State(empty_x, empty_y, goal_x, goal_y, 0)
    q.put(start)
    visited.add(start)

    dx = [0, 1, 0, -1]
    dy = [-1, 0, 1, 0]

    while not q.empty():
        current = q.get()

        if current.goal_x == 0 and current.goal_y == 0:
            return current.steps
        
        for direction in range(4):
            new_empty_x = current.empty_x + dx[direction]
            new_empty_y = current.empty_y + dy[direction]

            if (
                0 <= new_empty_x <= max_x and
                0 <= new_empty_y <= max_y and
                viable_nodes[new_empty_y][new_empty_x]
            ):
                new_goal_x, new_goal_y = current.goal_x, current.goal_y

                if new_empty_x == current.goal_x and new_empty_y == current.goal_y:
                    new_goal_x, new_goal_y = current.empty_x, current.empty_y

                new_state = State(new_empty_x, new_empty_y, new_goal_x, new_goal_y, current.steps + 1)

                if new_state not in visited:
                    visited.add(new_state)
                    q.put(new_state)

    return -1

nodes: list[Node] = []

# Skip the first two lines
input()
input()

pattern = r'node-x(\d+)-y(\d+)\s+(\d+)T\s+(\d+)T\s+(\d+)T\s+\d+%'

for line in sys.stdin:
    matches = re.search(pattern, line)
    if matches:
        x = int(matches.group(1))
        y = int(matches.group(2))
        size = int(matches.group(3))
        used = int(matches.group(4))
        avail = int(matches.group(5))

        nodes.append(Node(x, y, size, used, avail))
    
viable_pairs = 0
for i in range(len(nodes)):
    for j in range(len(nodes)):
        if i != j and nodes[i].used > 0 and nodes[i].used <= nodes[j].avail:
            viable_pairs += 1

print(f'Number of viable pairs: {viable_pairs}')

min_steps = solve_part2(nodes)
print(f'Minimum steps required: {min_steps}')

File: 2016/Day22/main.dart
import 'dart:collection';
import 'dart:io';

class Node {
  final int x, y;
  final int size, used, avail;

  Node(this.x, this.y, this.size, this.used, this.avail);
}

class State {
  final int emptyX, emptyY;
  final int goalX, goalY;
  final int steps;

  State(this.emptyX, this.emptyY, this.goalX, this.goalY, this.steps);

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;
    return other is State &&
        emptyX == other.emptyX &&
        emptyY == other.emptyY &&
        goalX == other.goalX &&
        goalY == other.goalY;
  }

  @override
  int get hashCode => Object.hash(emptyX, emptyY, goalX, goalY);
}

void printGrid(
  List<Node> nodes,
  int maxX,
  int maxY,
  int emptyX,
  int emptyY,
  int goalX,
) {
  final grid = List.generate(maxY + 1, (_) => List.filled(maxX + 1, '.'));

  for (final node in nodes) {
    if (node.used > 100) grid[node.y][node.x] = '#';
  }

  grid[emptyY][emptyX] = '_';
  grid[0][0] = 'O';
  grid[0][goalX] = 'G';

  print('Grid visualization:');
  for (int y = 0; y <= maxY; y++) {
    print(grid[y].join(' '));
  }
}

int solvePart2(List<Node> nodes) {
  int maxX = 0, maxY = 0;
  for (final node in nodes) {
    maxX = maxX > node.x ? maxX : node.x;
    maxY = maxY > node.y ? maxY : node.y;
  }

  int emptyX = -1, emptyY = -1;
  for (final node in nodes) {
    if (node.used == 0) {
      emptyX = node.x;
      emptyY = node.y;
      break;
    }
  }

  int goalX = maxX;
  int goalY = 0;

  final viableNodes = List.generate(
    maxY + 1,
    (_) => List.filled(maxX + 1, true),
  );

  for (final node in nodes) {
    if (node.used > 100) viableNodes[node.y][node.x] = false;
  }

  printGrid(nodes, maxX, maxY, emptyX, emptyY, goalX);

  final q = Queue<State>();
  final visited = <State>{};

  final start = State(emptyX, emptyY, goalX, goalY, 0);
  q.add(start);
  visited.add(start);

  const dx = [0, 1, 0, -1];
  const dy = [-1, 0, 1, 0];

  while (q.isNotEmpty) {
    final current = q.removeFirst();

    if (current.goalX == 0 && current.goalY == 0) return current.steps;

    for (int dir = 0; dir < 4; dir++) {
      final newEmptyX = current.emptyX + dx[dir];
      final newEmptyY = current.emptyY + dy[dir];

      if (newEmptyX >= 0 &&
          newEmptyX <= maxX &&
          newEmptyY >= 0 &&
          newEmptyY <= maxY &&
          viableNodes[newEmptyY][newEmptyX]) {
        int newGoalX = current.goalX;
        int newGoalY = current.goalY;

        if (newEmptyX == current.goalX && newEmptyY == current.goalY) {
          newGoalX = current.emptyX;
          newGoalY = current.emptyY;
        }

        final newState = State(
          newEmptyX,
          newEmptyY,
          newGoalX,
          newGoalY,
          current.steps + 1,
        );

        if (!visited.contains(newState)) {
          visited.add(newState);
          q.add(newState);
        }
      }
    }
  }

  return -1;
}

void main() {
  final nodes = <Node>[];

  // Skip the first two lines of input
  stdin.readLineSync();
  stdin.readByteSync();

  final pattern = RegExp(
    r'node-x(\d+)-y(\d+)\s+(\d+)T\s+(\d+)T\s+(\d+)T\s+\d+%',
  );

  String? line;
  while ((line = stdin.readLineSync()) != null && line!.isNotEmpty) {
    final match = pattern.firstMatch(line);
    if (match != null) {
      final x = int.parse(match.group(1)!);
      final y = int.parse(match.group(2)!);
      final size = int.parse(match.group(3)!);
      final used = int.parse(match.group(4)!);
      final avail = int.parse(match.group(5)!);

      nodes.add(Node(x, y, size, used, avail));
    }
  }

  int viablePairs = 0;
  for (int i = 0; i < nodes.length; i++) {
    for (int j = 0; j < nodes.length; j++) {
      if (i != j && nodes[i].used > 0 && nodes[i].used <= nodes[j].avail) {
        viablePairs++;
      }
    }
  }

  print('Number of viable pairs: $viablePairs');

  final minSteps = solvePart2(nodes);
  print('Minium steps required: $minSteps');
}
